hspec-core/src/Test/Hspec/Core/Runner/Eval.hs

```diff
@@ -53,22 +53,27 @@ every seconds action = do
 
 type FormatResult = Either E.SomeException Result -> FormatM ()
 
-parallelize :: Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-parallelize isParallelizable e
-  | isParallelizable = runParallel e
-  | otherwise = runSequentially e
+parallelize :: Maybe QSem -> Bool -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+parallelize jobsSem isParallelizable e
+  | isParallelizable = runParallel jobsSem e
+  | otherwise = runSequentially jobsSem e
 
-runSequentially :: (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
-runSequentially e reportProgress formatResult = return $ do
-  result <- liftIO $ evalExample (e reportProgress)
+limitParallelism :: Maybe QSem -> IO a -> IO a
+limitParallelism jobsSem = case jobsSem of
+  Nothing -> id
+  Just s -> E.bracket_ (waitQSem s) (signalQSem s)
+
+runSequentially :: Maybe QSem -> (ProgressCallback -> IO Result) -> ProgressCallback -> FormatResult -> IO (FormatM ())
+runSequentially jobsSem e reportProgress formatResult = return $ do
+  result <- liftIO $ limitParallelism jobsSem $ evalExample (e reportProgress)
```

Comment:
This is a little bit tough to explain, but we (@sol and I) were just discussing parallel execution of specs. I think that using `limitParallelism` here can result in all parallelizable examples to be executed first while the sequential ones are blocked. After the parallelizable examples are done the sequential ones are being executed in sequence. This is less than ideal parallelism. Ideally we would continue to run the sequential examples in the main thread while the parallelizable ones are also being run.

Maybe it's best to just remove `limitParallelism` here.

(Strictly speaking we may get one more thread than specified with `-j` in the presence of sequential examples. But we assume that's ok.)